### ISSUE TITLE
Demangled OpenMP parallel procedure has the same name as the procedure itself

### ISSUE THREAD

**Issue by laksono** [HPCToolkit/hpctoolkit#122]
Compiler: g++ 6.4.1
Application: Lulesh
Platform: ori (x86_64)

Demangling OpenMP parallel loops (which are translated to OpenMP parallel procedure) has the same name as the procedure itself.
For instance, in procedure `CalcEnergyForElems`, it has 5 parallel loops. The demangled name of these loops is also `CalcEnergyForElems` which is confusing. Furthermore, hpcviewer bottom-up view assumes that all procedure has different name, and it doesn't know to distinguish between the real procedure and the OpenMP parallel procedure since they have the same name.

I propose hpcprof to detect if a procedure has the same demangled name, but different mangled name, it has to append it with something like a line number (if available) to distinguish between different parallel loops.

![image](https://user-images.githubusercontent.com/8781032/55029415-02cc8980-4fd8-11e9-844e-35bb0ba72103.png)

Output from gdb shows that all OpenMP parallel regions in g++ 6.4.1 has the same demangled name (`CalcEnergyForElems`) if they are in the same function, but they have different mangled names (`_ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.XX`):
```
$56 = 0xd5d970 "CalcEnergyForElems"
$58 = 0x10fe530"_ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.19

$60 = 0x1654d60 "CalcEnergyForElems"
$62 = 0x148a160 "_ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.20"

$64 = 0x1321250 "CalcEnergyForElems"
$66 = 0xf4d5e0 "_ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.21"

$68 = 0xd422d0 "CalcEnergyForElems"
$70 = 0x1138ad0 "_ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.22

$72 = 0xd41450 "CalcEnergyForElems"
$74 = 0xdfd770 "_ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.23"
```

**Comment by laksono** [comment]
Problem in **hpcviewer** when clicking "show hot path" on flat view and bottom-up view, due to the same procedure name (I guess). This problem doesn't occur in top-down view.

```
!ENTRY org.eclipse.ui 4 0 2019-03-26 15:24:36.420
!MESSAGE Unhandled event loop exception
!STACK 0
org.eclipse.core.runtime.AssertionFailedException: null argument:
        at org.eclipse.core.runtime.Assert.isNotNull(Assert.java:85)
        at org.eclipse.core.runtime.Assert.isNotNull(Assert.java:73)
        at org.eclipse.jface.viewers.AbstractTreeViewer.getTreePathFromItem(AbstractTreeViewer.java:2925)
        at org.eclipse.jface.viewers.AbstractTreeViewer.internalGetWidgetToSelect(AbstractTreeViewer.java:1730)
        at org.eclipse.jface.viewers.AbstractTreeViewer.internalExpand(AbstractTreeViewer.java:1625)
        at org.eclipse.jface.viewers.AbstractTreeViewer.setSelectionToWidget(AbstractTreeViewer.java:2519)
        at org.eclipse.jface.viewers.StructuredViewer.setSelectionToWidget(StructuredViewer.java:1770)
        at org.eclipse.jface.viewers.AbstractTreeViewer.setSelectionToWidget(AbstractTreeViewer.java:2992)
        at org.eclipse.jface.viewers.StructuredViewer.setSelection(StructuredViewer.java:1726)
        at org.eclipse.jface.viewers.TreeViewer.setSelection(TreeViewer.java:1077)
        at edu.rice.cs.hpc.viewer.scope.ScopeViewActions.showHotCallPath(ScopeViewActions.java:298)
        at edu.rice.cs.hpc.viewer.scope.ScopeViewActionsGUI$3.widgetSelected(ScopeViewActionsGUI.java:488)
        at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:249)
        at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
        at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5227)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1340)
        at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4561)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4151)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1121)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
        at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:150)
        at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:693)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
        at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:610)
        at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
        at edu.rice.cs.hpc.viewer.framework.HPCViewer.start(HPCViewer.java:28)
        at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
        at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
        at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
        at org.eclipse.equinox.launcher.Main.main(Main.java:1492)
```

**Comment by laksono** [comment]
@jmellorcrummey : I assume this is not a problem with other compilers (like clang).

**Comment by jmellorcrummey** [comment]
It is important to handle this problem. In several cases, we have seen the demangler omit a suffix like _omp_fn.19. I think this issue occurs only for C++. I think that the shortened names without the suffix come from the demangler. I am not sure whether it is the demangler Dyninst or in binutils for this particular case. I know that the demangler in Dyninst has this problem because Marty and I have seen that the link name (ln) in an hpcstruct file contains the suffix but the name (n) is just the shortened version without the suffix.

@mwkrentel please investigate.

While you are at it, another demangler problem we have discussed before is present on arm and power but not x86_64 with hpcstruct using Dyninst 10.0.

In the following test. Compile this with 
gcc -o demangle demangle.c -fopenmp

In hpcstruct output, the procedure with link name "a" (ln="a") ends up having demangled name "signed char" (n="signed char") on arm1 and on po. This happens with hpcstruct built with Dyninst 10.0.

On arm1, I used the default gcc 7.3.1
On po, I used gcc 6.4.0
It gets the right result on  gpu.cs using gcc 7.3.0


```

#include <omp.h>
#include <stdio.h>

#define N 40
#define T 4 

void
loop(){
        int j;
        for(j=0;j<1000000000;j+=2) j--;
}

void
a()
{
        #pragma omp parallel num_threads(T)
        {
                loop();
                printf("Outer region: ID = %d\n", omp_get_thread_num());
        }
}

int
main()
{
  a();
  return 0;
}
```

**Comment by mwkrentel** [comment]
The short answer from hpcstruct's view is GIGO (no offense).

I searched in Laksono's directory on ori.cs for a lulesh binary and
found `lulesh-RAJA-parallel.exe` in `/home/la5/codes/test/perf_test/app/`.
I've put a copy in my home directory on mic.cs.

I found two functions at `0x406dd0` and `0x409420` which illustrate
the problem.

Both nm and objdump report the mangled and demangled names for
`0x406dd0` as:

```
0000000000406dd0 t _ZN4RAJA6forallIZ18CalcEnergyForElems ... SetET_._omp_fn.43

0000000000406dd0 <void RAJA::forall<CalcEnergyForElems(double*, ...
    ::{lambda(int)#1}) [clone ._omp_fn.43]>:
```

And `0x409420` as:

```
0000000000409420 t _ZN4RAJA6forallIZ18CalcEnergyForElems ... SetET_._omp_fn.42

0000000000409420 <void RAJA::forall<CalcEnergyForElems(double*, ...
    ::{lambda(int)#1}) [clone ._omp_fn.42]>:
```

I also tried the g++ builtin `__cxa_demangle()` and libiberty
`cplus_demangle()` functions directly.  Same answers as nm and objdump.

So, the correct answer is there in the binary.  Both the mangled and
demangled names are different.  But the answer from Symtab doesn't
have `[clone ._omp_fn.43]` making the pretty names look the same.

@mxz297 does that seem like what's happening?
Is Symtab dropping `[clone ._omp_fn.43]` for some reason?
I think that's what struct gets from `pretty_names_begin()`.

Note: hpcstruct never does any demangling on its own (unless you're
using the `--demangle-library` and `--demangle-function` options).
All of the demangling comes from Symtab.

Also note: although it's easy to call `cplus_demangle`, the hard part
is deciding **when** to demangle.  Not every name is mangled and
demangling a non-mangled name gives the wrong answer.  (I think this
is what's happening with the `a` --> `signed char` problem.)

What you have to do, and I assume what Symtab already does, is for
every symbol, see what module (CU) it comes from and what the source
language is.  Demangle if the language is C++.

So, no I don't think that adding a line number to the name is the
right solution.  Instead, we should first confirm what Symtab is
doing, if it's chopping the name, and if so, then fix that.

Btw, comparing g++ `__cxa_demangle` with libiberty `cplus_demangle`,
`cplus_demangle` has several options for how to demangle.  But
`__cxa_demangle` has no options, you get one answer, no way to
customize it.

In particular, with `cplus_demangle`, you can demangle type names or
not.  So, `a` is not a valid function name encoding, it's a type
encoding, `signed char`.  With the type names option turned off, then
`cplus_demangle` reports that `a` is not a mangled name.  But
`__cxa_demangle` always returns `signed char`.  Maybe cxxabi has some
other way to make this distinction, I don't know.

So, maybe Dyninst should look at whether it's better to use
`cplus_demangle` over `__cxa_demangle`.

**Comment by laksono** [comment]
I was using the binary on ori from different directory:

```/home/la5/git/hpctoolkit-tests/applications/lulesh/lulesh2.0```

**Comment by mwkrentel** [comment]
Same problem.  In the lulesh2.0 binary, there are 5 procs for openmp
loops in `CalcEnergyForElems`, for example:

```
<P i="4236"
   n="CalcEnergyForElems"
   ln="_ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.19"
   l="2091"
   v="{[0x408950-0x408951)}">
```

For all 5, Symtab provides the mangled name as above and the pretty
name as `CalcEnergyForElems`.

I don't understand why Symtab would demangle this name as the short
form like this.  Either this is a bug in Symtab (maybe some odd corner
case), or else there's something going on here that I don't understand.

We need to check with @mxz297.

**Comment by mxz297** [comment]
@mwkrentel @jmellorcrummey @laksono 

I feel there is some abstraction gap here between what dyninst provides and what you wants. For example, a SymtabAPI::Symbol can provide the following three types of names:

```
getMangledName(): _ZL18CalcEnergyForElemsPdS_S_S_S_S_S_S_S_S_S_S_S_dddddS_S_ddiPi._omp_fn.19

getPrettyName():CalcEnergyForElems

getTypedName():CalcEnergyForElems(double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double, double, double, double, double, double*, double*, double, double, int, int*) [clone ._omp_fn.19]
```

Based on your descriptions, hpcstruct seems to use information from `getPrettyName()`. 

In my impression, `getPrettyName()` should give what the programmer writes in the source code.  `getTypedName()` provides the complete demangled name, which will include type information and whatever other information expressed in the mangled name (in this case, it contains `[clone ._omp_fn.19]`). 

The abstraction gap is that openmp parallel procedures are not really individual source functions. At this point, I feel SymtabAPI is doing what it is designed to do. However, it is not doing what you want.

Are you good with complete demangled name? If you are, then you can switch to use `getTypedName()`. If not, what exactly do you want? Do you just want exclude types?

**Comment by jmellorcrummey** [comment]
We will test with getTypedName()

**Comment by jmellorcrummey** [comment]
Moving today's discussion into the issue from email ...

At 2:32pm @mwkrentel said:

First, let me check, what method does Symtab use to demangle names?
Is it the C++ __cxa_demangle() function?
Does Symtab make any adjustments to the answer it gets on its own?

I think we should first settle on what Symtab wants to produce with
mangled, pretty and typed names.  And I think Dyninst should decide
this based on what makes sense in general, not something specific to
hpctoolkit's needs.

Specifically, how much of the type annotations are supposed to be in
the pretty name?  I bring this up because I think we're getting
inconsistent answers from Symtab.

For example, in the first lulesh-RAJA-parallel.exe binary that I
looked at, there's one function, SumElemFaceNormal() at 0x40b450 with
mangled, pretty and typed names:

 _Z17SumElemFaceNormalPdS_S_S_S_S_S_S_S_S_S_S_dddddddddddd

 SumElemFaceNormal

 SumElemFaceNormal(double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double, double, double, double, double, double, double, double, double, double, double, double)

Fine.  But in the same binary, CalcEnergyForElems() looks like:
_ZN4RAJA6forallIZ18CalcEnergyForElemsPdS1_S1_S1_S1_S1_S1_S1_S1_S1_S1_S1_S1_dddddS1_S1_S1_ddPNS_8IndexSetEEUliE_EEvNS_21omp_parallel_for_execERKNS_9RangeISetET_._omp_fn.42

 RAJA::forall<CalcEnergyForElems(double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double, double, double, double, double, double*, double*, double*, double, double, RAJA::IndexSet*)::{lambda(int)#1}>

 void RAJA::forall<CalcEnergyForElems(double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double, double, double, double, double, double*, double*, double*, double, double, RAJA::IndexSet*)::{lambda(int)#1}>(RAJA::omp_parallel_for_exec, RAJA::RangeISet const&, CalcEnergyForElems(double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double, double, double, double, double, double*, double*, double*, double, double, RAJA::IndexSet*)::{lambda(int)#1}) [clone ._omp_fn.42]

Here, even the pretty name has lots of type info, and almost what we
want.  It's just missing '_omp_fn.42' to make it unique.

Is the difference here that one is a function and the other is a
template?  Do these demangle differently?

Maybe that's what is confusing me.

**Comment by jmellorcrummey** [comment]
At 2:54  @mxz297 said:

```
First, let me check, what method does Symtab use to demangle names?
Is it the C++ __cxa_demangle() function?
Does Symtab make any adjustments to the answer it gets on its own?
```

Symtab calls __cxa_demangle to demangle names. Based on the code in
Dyninst, getTypedName() returns what __cxa_demangle returns.

getPrettyName() takes what __cxa_demangle returns and processes it
with a function "dedemangle". This "dedemangle" function is committed
by Bill in 2010.

The code comment at the beginning of this function explains what it
mechanically does (dyninst/common/src/string-regex.C)

  /* Lifted from Jeffrey Odom.  Code reformatted so
     I could figure out how to eliminate compiler warnings.
     Adjusted to handle spaces inside templates intelligently.
     We cut off everything after the first l-paren, so we don't
     need to worry about the space after the parameters but
     before the 'const'. */

The sentence "we cut off everything after the first l-paren" explains
why getPreetyName() will strip the "_omp_fn.42" suffix. The
implementation of this function suggests that it will try to keep c++
template types.

So, it looks like Bill treated template type information as part of
source function name.


```
I think we should first settle on what Symtab wants to produce with
mangled, pretty and typed names.  And I think Dyninst should decide
this based on what makes sense in general, not something specific to
hpctoolkit's needs.

Specifically, how much of the type annotations are supposed to be in
the pretty name?  I bring this up because I think we're getting
inconsistent answers from Symtab.

```
If you agree that template type information is part of the source
function name, then you are not getting inconsistent answers from
Symtab.


```
For example, in the first lulesh-RAJA-parallel.exe binary that I
looked at, there's one function, SumElemFaceNormal() at 0x40b450 with
mangled, pretty and typed names:

  _Z17SumElemFaceNormalPdS_S_S_S_S_S_S_S_S_S_S_dddddddddddd

  SumElemFaceNormal

  SumElemFaceNormal(double*, double*, double*, double*, double*,
double*, double*, double*, double*, double*, double*, double*, double,
double, double, double, double, double, double, double, double, double,
double, double)

Fine.  But in the same binary, CalcEnergyForElems() looks like:
_ZN4RAJA6forallIZ18CalcEnergyForElemsPdS1_S1_S1_S1_S1_S1_S1_S1_S1_S1_S1_S1_dddddS1_S1_S1_ddPNS_8IndexSetEEUliE_EEvNS_21omp_parallel_for_execERKNS_9RangeISetET_._omp_fn.42

  RAJA::forall<CalcEnergyForElems(double*, double*, double*, double*,
double*, double*, double*, double*, double*, double*, double*, double*,
double*, double, double, double, double, double, double*, double*,
double*, double, double, RAJA::IndexSet*)::{lambda(int)#1}>

  void RAJA::forall<CalcEnergyForElems(double*, double*, double*,
double*, double*, double*, double*, double*, double*, double*, double*,
double*, double*, double, double, double, double, double, double*,
double*, double*, double, double,
RAJA::IndexSet*)::{lambda(int)#1}>(RAJA::omp_parallel_for_exec,
RAJA::RangeISet const&, CalcEnergyForElems(double*, double*, double*,
double*, double*, double*, double*, double*, double*, double*, double*,
double*, double*, double, double, double, double, double, double*,
double*, double*, double, double, RAJA::IndexSet*)::{lambda(int)#1})
[clone ._omp_fn.42]

Here, even the pretty name has lots of type info, and almost what we
want.  It's just missing '_omp_fn.42' to make it unique.

Is the difference here that one is a function and the other is a
template?  Do these demangle differently?

```
I personally don't like the "getPrettyName()" interface, but I don't
have the intent to change it yet. It is not clear to me what's the
definition of pretty... Obviously, based on the code: pretty =
template types + source function name.

Thanks,

--Xiaozhu

**Comment by jmellorcrummey** [comment]
Adding @martyitz to this thread (follow the link in the email to view the whole conversation on GitHub)

At 4:04pm @mwkrentel said:

Ok, I think that explains my confusion.
Thanks!

When I looked at this 3-4 years ago, I decided that pretty was better
than typed for our purposes.  We haven't really had a problem until
now, but I must have been looking at some skewed examples.

I'll do some more experiments, but I think I'm inclined to switch to
typed.  Full typed names are grossly long in some cases, but I think
they may be necessary.

As for pretty name, I agree that I'm not sure what it should mean, or
even if it should exist at all.  The take away is that pretty name is
an internal Symtab adjustment or smoothing, not an external standard
with a more clear right or wrong.  (So, maybe Dyninst apps might want
to do their own smoothing starting from the typed name.)

And let me check, you do test if the source language is C++ and only
demangle if it is, right?  That is, you don't just always demangle every
name.

Btw, I think you may want to switch from __cxa_demangle() to the
libiberty cplus_demangle().  The libiberty function has more options
for how to demangle, and I've seen examples where this matters
(eg, a function name versus a type encoding).

Also, I think I've seen a couple examples where Symtab demangles
something twice and thus produces a strange answer.

But that's just a heads up that I think I've seen some odd things.
I'll let you know when I have something more specific.

**Comment by mwkrentel** [comment]
@laksono Is there something in the viewer that is matching up procs by
their pretty names?

If yes, then why is it using pretty names instead of link names?

I'm inclined to change our use to typed names.  But I'm not sure if
it's correct for the viewer to treat the pretty names as anything more
than a decoration.

**Comment by martyitz** [comment]
Can you make it a switch, so we can see how the alternatives look?

PA had a viewer switch between long C++ names, short names (missing the argument types), and
mangled names, with or without the .so name appended.

	Marty

**Comment by laksono** [comment]
> @laksono Is there something in the viewer that is matching up procs by
> their pretty names?

No, there isn't.
hpcprof outputs only a single name of proc, and the viewer will get this name in the database.

> 
> If yes, then why is it using pretty names instead of link names?
> 
> I'm inclined to change our use to typed names. But I'm not sure if
> it's correct for the viewer to treat the pretty names as anything more
> than a decoration.

**Comment by mxz297** [comment]
> And let me check, you do test if the source language is C++ and only
> demangle if it is, right? That is, you don't just always demangle every
> name.

I am not 100% about my conclusion, but based on the Dyninst code I have looked at, it looks like SymtabAPI will demangle every name. Do you think this will be a 10.1 thing or not?

> 
> Btw, I think you may want to switch from __cxa_demangle() to the
> libiberty cplus_demangle(). The libiberty function has more options
> for how to demangle, and I've seen examples where this matters
> (eg, a function name versus a type encoding).
> 

I have little knowledge about the difference between these two demanglers. Do you think this has to be a 10.1 thing or can be later?

> Also, I think I've seen a couple examples where Symtab demangles
> something twice and thus produces a strange answer.
> 
> But that's just a heads up that I think I've seen some odd things.
> I'll let you know when I have something more specific.

**Comment by martyitz** [comment]
I think that a new demangle algorithm for RAJA codes is needed.  Clearly, the
long string of argument types is meaningless for the user.  Perhaps the functions
should be named by the source file and line where they are created.  It needs to
be something that conveys to the user where to look in his or her source.

**Comment by jmellorcrummey** [comment]
The hyperlink from the function in hpcviewer's scope pane will take you to the source file and line number where the function is created. However, one can compile without line map information and not have source file and line information. Providing the detangled name is the best we can do in this case, even though it is not very meaningful.

**Comment by martyitz** [comment]
> On Apr 2, 2019, at 8:58 AM, John Mellor-Crummey <notifications@github.com> wrote:
> 
> The hyperlink from the function in hpcviewer's scope pane will take you to the source file and line number where the function is created. However, one can compile without line map information and not have source file and line information. Providing the detangled name is the best we can do in this case, even though it is not very meaningful.
> 
 
I assume “detangled” was an intentional pun.  :-)

How could the user figure out what the demangled name refers to?  I’d use the mangled name in this case, perhaps with a
[compile with -g, you dummy] tag appended.

**Comment by mwkrentel** [comment]
> I am not 100% about my conclusion, but based on the Dyninst code I
> have looked at, it looks like SymtabAPI will demangle every name. Do
> you think this will be a 10.1 thing or not?

Unfortunately, I think Symtab is demangling every name.  John's simple
openmp example above shows the problem.  If you compile it as a C
program (not C++), then Symtab reports the mangled name as 'a' and the
pretty and typed names as 'signed char'.

Fortunately, the fix is small and simple.  Symtab already computes the
source language per module.  If the source language is
`lang_CPlusPlus` or `lang_GnuCPlusPlus`, then demangle the name,
otherwise don't.

I'm not sure what `lang_Fortran_with_pretty_debug` means.  But there
are no options in libiberty `demangle.h` related to Fortran, so I'm
guessing 'pretty debug' doesn't mean demangle here.


> I have little knowledge about the difference between these two
> demanglers. Do you think this has to be a 10.1 thing or can be later?

This deserves more study.  The libiberty version has useful options
that `__cxa_demangle` doesn't have.  But I'd wait until I have a clear
example of something that `__cxa_demangle` just can't do correctly.
Plus, fix the previous problem first.

**Comment by mwkrentel** [comment]
I've mostly finished my investigation on this.  I've reviewed all the
places where we get function names, how we get them, and if and how we
demangle them.  There are two things to do.

1. I'll change the way we get the mangled name from Symtab from the
"pretty" name to typed name.  That should make the demangled names
unique and solve Laksono's original problem.

2. We'll ask @mxz297 to fix Symtab so that it doesn't automatically
demangle every symbol.  Instead, it should lookup the module for the
symbol and demangle if and only if the source language is C++ or GNU
C++.  That should fix the 'a' --> 'signed char' problem.

I think all the infrastructure for that is in place, so hopefully this
is a fairly small fix and can go into 10.1.  I'll open a dyninst issue
for this.

I also looked at the difference between `__cxa_demangle()` and
libiberty `cplus_demangle()`.  And now, I don't see any need to switch
to the libiberty version.

The main difference is that `cplus_demangle("a")` reports that "a" is
not a valid function name.  (It's a type encoding, not a function
mangling.)  But `__cxa_demangle("a")` always returns `signed char`, 
which is technically correct, but in the wrong context.

So, the real problem is deciding to demangle "a" in the first place.
The libiberty version would hide that sin, but as long as you only
demangle symbol names from C++ source files, then I don't see any real
advantage in using libiberty.

Btw, my previous claim that hpcstruct doesn't demangle anything on its
own is not correct.  That's true for the primary case, where we have a
valid Symtab function, by far the most common case.

But there are two other cases: outlined openmp functions and plt
functions where there is no Symtab name so we construct one.  Second,
inline names.  In that case, we get a (file, line, proc) triple from
Symtab, but it's only one proc name, normally mangled, but actually
whatever is in the dwarf info.  In this case, we look up the source
file and demangle if it comes from C++.
